### ISSUE TITLE
Make EOED transmission optional in QUIC, please

### ISSUE THREAD

**Issue by huitema** [quicwg/base-drafts#1518]
The TLS spec seems to imply that the End Of Early Data (EOED) message, if required, should be protected and transmitted as 0-RTT data.  I found that when implementing the draft-13 spec, and it was somewhat of a surprise. I understand the willingness to exactly parallel TLS 1.3 behavior of TCP, but it is weird when packets are sent as UDP datagrams.

The weirdest part is probably the retransmission behavior. 0-RTT packets are only acked by 1-RTT packets. Stream frames in 0-RTT packets are resent in 1-RTT packets. For correct behavior, we must guarantee that it is received before the Client Finished message. What are implementers supposed to do? Put a huge wart in the retransmission control?

It would be much simpler to not require an EOED message for Quic. EOED is not really functional, since 0-RTT packets are not processed through the TLS stack. 0-RTT and 1-RTT packets are in the same sequence space, there is no need for a demarcation. It is just one of those hazing rituals that are imposed on developers.

**Comment by ekr** [comment]
On Tue, Jul 3, 2018 at 8:16 PM, Christian Huitema <notifications@github.com>
wrote:

> The TLS spec seems to imply that the End Of Early Data (EOED) message, if
> required, should be protected and transmitted as 0-RTT data.
>

Yes. All messages are transmitted in the same epochs they would be in TLS
1.3.

I found that when implementing the draft-13 spec, and it was somewhat of a
> surprise. I understand the willingness to exactly parallel TLS 1.3 behavior
> of TCP, but it is weird when packets are sent as UDP datagrams.
>
> The weirdest part is probably the retransmission behavior. 0-RTT packets
> are only acked by 1-RTT packets. Stream frames in 0-RTT packets are resent
> in 1-RTT packets. For correct behavior, we must guarantee that it is
> received before the Client Finished message. What are implementers supposed
> to do? Put a huge wart in the retransmission control?
>
Well, EOED isn't a stream frame, so I don't think it's actually that
complicated to implement this. I think it's more important to keep the TLS
1.3 state machine intact.

-Ekr

**Comment by huitema** [comment]
It really depends how you deal with retransmissions. In my code it is a huge pain. Of course I can find some contorted way to deal with it. But I hate that kind of stuff.

As for the TLS 1.3 state machine, what does the EOED actually achieve? You cannot clear the 0-RTT encryption context until you are sure that all 0-RTT datagrams have been received. Do you just want to prove that the client could exercise the 0-RTT key?

**Comment by martinthomson** [comment]
It says "handshake data", which is correct, if perhaps a little too easy to mistake for "Handshake packets".

Acknowledging promptly is primarily an optimization.  It should still work with the leisurely timers.  My naive assumption was that when you process a CRYPTO frame you would cut the ack delay timer.  That gets you a pretty good approximation of the right behaviour, even if that also means that NewSessionTicket forces more aggressive acknowledgment than necessary (for which you could condition on the packet protection level if you cared).

**Comment by ianswett** [comment]
I agree with Christian, this is a pain and an ugly special case, since everything else sent in 0RTT can be retransmitted in 1RTT.

I think this is a case where the stream 0 QUIC record layer design should make EOED unnecessary.

**Comment by kazuho** [comment]
I do not think that the requirement to send EOED using a 0-RTT packet is complex enough to justify the changes to the TLS handshake transcript.

My understanding is that TLS stacks emits EOED when 1-RTT write key becomes available (otherwise in TLS over TCP a client will have a moment when it cannot send any data, which is silly). 1-RTT read key will be available at that moment, so you can always be certain that you can receive an ACK for EOED.

In other words, retransmission is not an issue.

Note also that clients have the freedom to continue retransmitting 0-RTT packets with application data even after receiving the 1-RTT write key. That is what TLS over TCP exactly does.

**Comment by huitema** [comment]
@kazuho can you expand a bit the argument about "the changes to the TLS handshake transcript"? Are you concerned about reusing the TLS code as is, or about not invalidating the results of the TLS1.3 correctness proofs?

**Comment by kazuho** [comment]
@huitema The latter.

**Comment by ianswett** [comment]
My current thinking is that we should make it QUIC v2 and try to convince someone to spend the time to do the correctness proofs in that timeframe, so we don't need this forever, and possibly prove we don't need CFIN as well.  And whatever other warts we have discovered in this process.

**Comment by vasilvv** [comment]
I'm not sure I agree with the "correctness proof" argument.  EndOfEarlyData exists to protect against truncation attacks on 0-RTT data; as such, it's a part of record layer, not handshake layer.  In fact, it did not go into transcript before draft19, so most papers cited for handshake analysis in the RFC do not mention it (out of those that do, the INRIA/MSR analysis treats it as a record layer entity, and _TLS 1.3 in Tamarin_ explicitly states it does not model it at all).

Given that this is unnecessary and adds bunch of complex and ugly logic, I think we should remove this.

**Comment by davidben** [comment]
Prior to it being a handshake message, we had an end_of_early_data *alert* and a KeyUpdate *message*. This combination was kind of weird, and since other post-handshake messages exist, so the application record processor must already account for handshake message interrupts, making it a message was simpler. Thus it was made a message for consistency. Messages during the handshake go into the transcript, so in it went. That resulted in a happy accident in stopping servers from breaking streaming 0-RTT, so that all worked well. (Before the change, folks kept trying to read 0-RTT data and *then* decide whether to accept early data, without realizing that [is forbidden and deadlocks](https://tools.ietf.org/html/rfc8446#section-4.2.11.3). Putting it in the transcript orders early data processing after ServerHello, which makes the processing order implicitly clear.)

Neither of these considerations apply to QUIC, and then handshake transcript trivially remains injective with it removed, so removing it seems plausible, and it sounds like it's complicating things a lot. At the same time, QUIC should avoid unnecessary modifications to TLS to keep abstractions clear, so the bar for such tweaks must be high. Otherwise TLS and QUIC will not be able to evolve independently.

(Which of these two considerations is stronger in this case, I don't have particular opinions. It does not sound frivolous, but I don't know enough about QUIC to judge how un-frivolous it is.)

**Comment by huitema** [comment]
@davidben if I understand correctly, the main issue is then compatibility with the existing TLS implementations. They expect the EOED to be received before the first handshake message. 

For me the simplest way to achieve that without changing the TLS stack would be to carry the EOED data in the handshake packet rather than the 0-RTT packet, and then deliver it to the TLS stack while pretending that it was carried with 0-RTT encryption. The added cost is a new frame type, but then I could remove all the warts about retransmitting 0-RTT data sometimes as 1-RTT and sometimes as 0-RTT.

Even simpler of course would be a TLS API option to "not bother with EOED when running over a datagram transport". That would remove all the EOED warts in the QUIC code. It would add some of them back in the TLS code, but probably at a much lower cost.

**Comment by davidben** [comment]
Right, I'm assuming that, given the rest of the stream zero business, the TLS stack will need modifications anyway, at which point it is perfectly capable of the occasional tweak for QUIC.

I wouldn't want to have too many of these and obviously you especially shouldn't change something substantial that would break analysis or future TLS extensions, but EndOfEarlyData seems unlikely to matter much. (Famous last words...)

**Comment by ekr** [comment]
Just to follow up on this point, removing EOED does allow a weak extension attack against data in the 0-RTT epoch. Consider the case where the client (a) sends EOED  and then (b) is compromised. Under these conditions an attacker will be able to send data which will appear to the server to be in 0-RTT. As I say, this is quite a weak attack because (a) the client could just delete the 0-RTT keys and (b) the attacker would presumably be able to instead send 1-RTT data instead. But it is a change in the properties of the system.

**Comment by ianswett** [comment]
@ekr: I don't think "EOED tells you that you will not receive messages with 0-RTT keys but with greater PNs." is providing much, since if EOED is lost, the receiver will continue to process 0RTT packets with higher packet numbers.

Also, you have to continue processing these 0RTT packets long after EOED has been initially received, because the sender of EOED may not know EOED has been received.

I'd argue in QUIC EOED is actually telling you very little, and I tend to think it's inclusion is more harmful than helpful.  It requires keeping 0RTT keys around and continue processing them for longer.  Without EOED, you can stop processing 0RTT packets once a single packet is received with 1RTT encryption, which seems like a much stronger signal.

**Comment by ekr** [comment]
1. Yes, the receiver will continue to process them if EOED is lost. That's why you retransmit them.
2. You don't have to continue processing higher-numbered 0-RTT packets after EOED is received because the client will never send them.
3. You can't stop processing 0RTT packets once a single packet is received in epoch 3 unless you want gaps due to reordering.

**Comment by huitema** [comment]
We do not have a rule in the transport draft that says that clients shall stop sending more 0-RTT packets after EOED has been sent. The separation between transport and TLS prevents that. The TLS stack at some points tells the client to "please forward these bytes on the epoch=1 crypto stream". The client does not know what the bytes mean. It has no concept of EOED.

Of course, we could introduce an extra synchronization between the transport and the TLS stack. The transport could tell the TLS stack that it has sent all the queued 0-RTT packets. Or the TLS stack could somehow signal to the transport that it should not send any more 0-RTT packets -- except of course for those 0-RTT packets required to send TLS messages, or to repeat them if no ACK arrives. I am definitely not suggesting that we do anything like that.

**Comment by ekr** [comment]
After EOED is sent, the TLS stack tells the transport stack that further non-TLS data should be sent with the 1-RTT keys at this point. How did you think that the transition happened other than that?

**Comment by ekr** [comment]
This is shown in Figure 3 of the TLS draft, though perhaps some more text is required.

**Comment by mikkelfj** [comment]
> 2. You don't have to continue processing higher-numbered 0-RTT packets after EOED is received because the client will never send them.

I don't get all the details of this but would you exepct higher numbered 0-RTT transmissions after EOED if there are retransmissions and EOED packet hasn't been acked.

**Comment by ianswett** [comment]
Re #2, it it possible the client misses an acknowledgement of the EOED.  If this happens, we either need to rely on implicit acknowledgement, which we've avoided so far, or the client retransmitting EOED and the server ACKing it.  Or never stop ACKing EOED until you know it's been received, but that seems more complex?

Re #3, I had higher numbered packets in mind, due to reordering.  But this seems very similar to how changing key phases would work, which matches my mental model of the 0RTT to 1RTT transition.

**Comment by ekr** [comment]
(2) Ian, sorry, yes, I meant just application data.
(3) Hmm... I was thinking these were separate sequence number spaces like in DTLS, but I guess they're the same, so maybe you're right. I'll have to think about that.

**Comment by huitema** [comment]
@ekr You write that "_After EOED is sent, the TLS stack tells the transport stack that further non-TLS data should be sent with the 1-RTT keys at this point_." I don't know for other stacks, but there is no such signal with Picotls. That API provides signals such as "expecting data from epoch X", "Here is data to send on epoch Y", "here is the write (or read) key for epoch X" and "all done". The signal "here is the write key for epoch 3" does not double as "and please stop sending on epoch 1". It cannot, because the epoch 1 crypto stream still has to be transmitted and acknowledged.

We could also have an interesting discussion about stream frames carried in the same 0-RTT packet as crypto frames. Should we have a different treatment depending on whether the stream frame is carried before or after the crypt frame in the packet?

**Comment by ekr** [comment]
Well, you certainly need some signal that says "send application data in epoch 3". What is that signal?

**Comment by huitema** [comment]
There is no explicit signal, but it is very much assumed that if the TLS stack provides the transport with the write key for epoch 3, then application data will be sent that way. I read again the QUIC/TLS draft, and I could not find a statement saying that application data MUST NOT be sent as 0-RTT if the 1-RTT key is available. But in fact, that's the whole point of this thread. If it were not for the transmission of EOED over 0-RTT, the transport could simply just discard the 0-RTT write keys when the 1-RTT write keys become available. I would very much like to be able to do that.

**Comment by vasilvv** [comment]
> I don't think the main issue is compatibility with the existing TLS implementations but rather (a) having consistency with TLS

The point I was trying to make is that EOED is not logically part of TLS handshake protection (which we're trying to preserve in order to rely on security properties proven in multiple analyses), but rather of TLS record layer.  We're pretty much rolling our own record layer, which means it befalls on us to ensure its security ourselves.  That does not itself mean we should drop EOED by itself, merely that we actually *have* to decide something without pointing fingers to TLS 1.3 :)

Regarding your attack, if you're concerned about 0-RTT keys being compromised after EOED, wouldn't you want to drop them as soon as possible?  As it stands, the client can't do it because it has to hold onto them for retransmitting EOED.
